tuna/example/example_lib.py

```diff
@@ -0,0 +1,149 @@
+"""Example library, integrated with MITuna, runs 'rocminfo' cmd"""
+
+import sys
+
+from tuna.mituna_interface import MITunaInterface
+from tuna.parse_args import TunaArgs, setup_arg_parser, clean_args, args_check
+from tuna.utils.miopen_utility import load_machines
+from tuna.libraries import Library
+from tuna.utils.db_utility import create_tables
+from tuna.example.example_tables import get_tables
+from tuna.example.example_worker import ExampleWorker
+from tuna.example.session import SessionExample
+
+
+class Example(MITunaInterface):
+  """Class to support an example of 'romcinfo' run"""
+
+  def __init__(self):
+    super().__init__(library=Library.EXAMPLE)
+    self.args = None
+
+  def parse_args(self):
+    # pylint: disable=too-many-statements
+    """Function to parse arguments"""
+    parser = setup_arg_parser('Example library integrated with MITuna', [
+        TunaArgs.ARCH, TunaArgs.NUM_CU, TunaArgs.VERSION, TunaArgs.SESSION_ID,
+        TunaArgs.MACHINES, TunaArgs.REMOTE_MACHINE, TunaArgs.LABEL,
+        TunaArgs.RESTART_MACHINE, TunaArgs.DOCKER_NAME
+    ])
+
+    parser.add_argument('--init_session',
+                        action='store_true',
+                        dest='init_session',
+                        help='Set up a new tuning session.')
+
+    clean_args('EXAMPLE', 'example')
+    args = parser.parse_args()
+    if len(sys.argv) == 1:
+      parser.print_help()
+      sys.exit(-1)
+
+    args_check(args, parser)
+
+    return args
+
+  def launch_worker(self, gpu_idx, f_vals, worker_lst, args):
+    """! Function to launch worker
+      @param gpu_idx Unique ID of the GPU
+      @param f_vals Dict containing runtime information
+      @param worker_lst List containing worker instances
+      @param args The command line arguments
+      @retturn ret Boolean value
+    """
+
+    kwargs = self.get_kwargs(gpu_idx, f_vals, args)
+    worker = ExampleWorker(**kwargs)
+    if args.init_session:
+      SessionExample().add_new_session(args, worker)
+      return False
+
+    worker.start()
+    worker_lst.append(worker)
+    return True
+
+  def compose_worker_list(self, res, args):
+    # pylint: disable=too-many-branches
+    """! Helper function to compose worker_list
+      @param res DB query return item containg available machines
+      @param args The command line arguments
+    """
+    worker_lst = []
+    for machine in res:
+      if args.restart_machine:
+        machine.restart_server(wait=False)
+        continue
+
+      #determine number of processes by compute capacity
+      # pylint: disable=duplicate-code
+      worker_ids = super().determine_num_procs(machine)
+      if len(worker_ids) == 0:
+        return None
+
+      f_vals = super().get_f_vals(machine, worker_ids, self.args)
+
+      for gpu_idx in worker_ids:
+        self.logger.info('launch mid %u, proc %u', machine.id, gpu_idx)
+        if not self.launch_worker(gpu_idx, f_vals, worker_lst, args):
+          break
```

Comment:
```suggestion
        if not self.launch_worker(gpu_idx, f_vals, worker_lst, args):
          raise ValueError('The worker failed to launch')
```

Reply:
do you want to break out of the outside loop as well? the inside loop goes over gpus and the outside goes over machines. Do you want to break both loops? do you want to handle the error here or let propagate?